classylss/cosmology.py

```diff
@@ -90,60 +152,117 @@ def __getattr__(self, name):
-        for key in list(pars.keys()):
-            if pars[key] is None: pars.pop(key)
+            if name in conflicts:
+                alias = aliases.get(name, name)
+                args[name] = getattr(toret, alias)
+            else:
+                for c in conflicts:
+                    if c == 'sigma8': continue
+                    if name in conflicts[c]:
+                        alias = aliases.get(c, c)
+                        args[c] = getattr(toret, alias)
 
-        assert args['gauge'] in ['synchronous', 'newtonian']
+        # set the gauge
+        args['gauge'] = 'newtonian' if toret.gauge == 0 else 'synchronous'
 
-        if 'P_k_max_h/Mpc' not in pars:
-            pars['P_k_max_h/Mpc'] = pars.pop('P_k_max')
+        # set sigma8 norm
+        if not any(par in pars for par in conflicts['sigma8']):
+            pars['A_s'] = toret.A_s
 
-        if 'z_max_pk' not in pars:
-            pars['z_max_pk'] = pars.pop('P_z_max')
+        toret.args = args
+        toret.extra = pars
 
-        for m in args['m_ncdm']:
-            if m == 0:
-                raise ValueError("A zero mass is specified in the massive neutrino list.")
+        return toret
 
-        # from CLASS notes:
-        # one more remark: if you have respectively 1,2,3 massive neutrinos,
-        # if you stick to the default value pm equal to 0.71611, designed to give m/omega of
-        # 93.14 eV, and if you want to use N_ur to get N_eff equal to 3.046 in the early universe,
-        # then you should pass here respectively 2.0328,1.0196,0.00641
+    def __setstate__(self, state):
 
-        N_ur_table = [3.046, 2.0328, 1.0196, 0.00641]
+        # remember for serialization
+        self.args, self.extra = state
 
-        if 'N_ur' not in pars:
-            pars['N_ur'] = N_ur_table[len(args['m_ncdm'])]
+        # remove sigma8 for norm later
+        desired_sigma8 = self.args.pop('sigma8', None)
 
-        pars['N_ncdm'] = len(args['m_ncdm'])
+        # verify and set defaults
+        pars = verify_parameters(self.args, self.extra)
 
         # initialize the engine as the backup delegate.
         self.engine = ClassEngine(pars)
-        self.delegates = {
-            ClassEngine: self.engine,
-        }
+        self.delegates = {ClassEngine: self.engine}
+
+        # set sigma8 by re-scaling A_s, if A_s was not specified
+        if desired_sigma8 is not None:
+            if not any(key in self.extra for key in conflicts['sigma8']):
+                self.sigma8 = desired_sigma8
 
     def clone(self, **kwargs):
-        """ create a new cosmology based on modification of self """
-        args = {}
-        args.update(self.args)
-        args.update(self.kwargs)
+        """
+        Create a new cosmology based on modification of self
+        """
+        desired_sigma8 = kwargs.pop('sigma8', None)
+        if desired_sigma8 is not None:
+            kwargs['A_s'] = self.A_s * (desired_sigma8/self.sigma8)**2
```

Comment:
It smells a bit strange the logic for handling sigma8 is scattered across a few functions. Is there a way to centralize them? Or it already is but just my impression was inaccurate?